tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/Test_ResizeAndComputePrimitives.cpp

```diff
@@ -0,0 +1,191 @@
+      evolution::dg::subcell::Tags::ActiveGrid, cons_tag, prim_tag,
+      gr::Tags::SpatialMetric<3, Frame::Inertial, DataVector>,
+      gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>,
+      gr::Tags::SqrtDetSpatialMetric<DataVector>, ::domain::Tags::Mesh<3>,
+      evolution::dg::subcell::Tags::Mesh<3>,
+      hydro::Tags::EquationOfState<
+          std::unique_ptr<EquationsOfState::EquationOfState<true, 1>>>>>(
+      active_grid, cons_vars, prim_vars, spatial_metric, inv_spatial_metric,
+      sqrt_det_spatial_metric, dg_mesh, subcell_mesh, std::move(eos));
+
+  using recovery_schemes = tmpl::list<
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::KastaunEtAl,
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::NewmanHamlin,
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::PalenzuelaEtAl>;
+  db::mutate_apply<grmhd::ValenciaDivClean::subcell::ResizeAndComputePrims<
+      recovery_schemes>>(make_not_null(&box));
+
+  REQUIRE(db::get<prim_tag>(box).number_of_grid_points() == active_num_pts);
+  if (active_grid == evolution::dg::subcell::ActiveGrid::Dg) {
+    prim_vars.initialize(cons_vars.number_of_grid_points());
+    grmhd::ValenciaDivClean::PrimitiveFromConservative<recovery_schemes>::apply(
+        make_not_null(
+            &get<hydro::Tags::RestMassDensity<DataVector>>(prim_vars)),
+        make_not_null(
+            &get<hydro::Tags::SpecificInternalEnergy<DataVector>>(prim_vars)),
+        make_not_null(
+            &get<hydro::Tags::SpatialVelocity<DataVector, 3>>(prim_vars)),
+        make_not_null(
+            &get<hydro::Tags::MagneticField<DataVector, 3>>(prim_vars)),
+        make_not_null(
+            &get<hydro::Tags::DivergenceCleaningField<DataVector>>(prim_vars)),
+        make_not_null(&get<hydro::Tags::LorentzFactor<DataVector>>(prim_vars)),
+        make_not_null(&get<hydro::Tags::Pressure<DataVector>>(prim_vars)),
+        make_not_null(
+            &get<hydro::Tags::SpecificEnthalpy<DataVector>>(prim_vars)),
+        get<grmhd::ValenciaDivClean::Tags::TildeD>(cons_vars),
+        get<grmhd::ValenciaDivClean::Tags::TildeTau>(cons_vars),
+        get<grmhd::ValenciaDivClean::Tags::TildeS<Frame::Inertial>>(cons_vars),
+        get<grmhd::ValenciaDivClean::Tags::TildeB<Frame::Inertial>>(cons_vars),
+        get<grmhd::ValenciaDivClean::Tags::TildePhi>(cons_vars), spatial_metric,
+        inv_spatial_metric, sqrt_det_spatial_metric,
+        db::get<hydro::Tags::EquationOfStateBase>(box));
+  }
+  CHECK_VARIABLES_APPROX(db::get<prim_tag>(box), prim_vars);
+}
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.ValenciaDivClean.Subcell.ResizeAndComputePrims",
+    "[Unit][Evolution]") {
+  MAKE_GENERATOR(gen);
+  // Use a small range of random values since we need recovery to succeed and we
+  // also reconstruct to the DG grid from the FD grid and need to maintain a
+  // somewhat reasonable state on both grids.
+  std::uniform_real_distribution<> dist(0.01, 0.010005);
```

Comment:
is this limit because you messed up the Lorentz factor?

Reply:
no, it's because reconstructing random numbers leads to wild oscillations

Reply:
how about a narrow range around 0.5?  that at least would have caught your incorrect lorentz factor

Reply:
seems to work so far. I'll squash it in once @fmahebert has looked at the fixups